src/main/java/com/cloudbees/jenkins/support/filter/WordsTrie.java

```diff
@@ -0,0 +1,171 @@
+        this.root = new TrieNode(false);
+    }
+
+    /**
+     * Add a word to the Trie.
+     * @param word the word
+     */
+    public void add(String word) {
+        TrieNode ref = root;
+        int i = 0;
+        while (i < word.length() - 1) {
+            // Fill in down the Trie
+            ref = ref.getOrCreate(word.charAt(i), s -> new TrieNode(false));
+            i++;
+        }
+        // We need to mark the last node as an end node
+        ref.getOrCreate(word.charAt(i), s -> new TrieNode(true)).end = true;
+    }
+
+    /**
+     * Get the regex String of this Trie.
+     *
+     * @return the regex String of this Trie.
+     */
+    public String getRegex() {
+        return root.getRegex();
+    }
+
+
+    static class TrieNode {
+
+        private final Map<Character, TrieNode> data = new TreeMap<>();
+
+        /**
+         * Mark this node as the end of a word
+         */
+        private boolean end;
+
+        TrieNode(boolean end) {
+            this.end = end;
+        }
+
+        /**
+         * Produce the regex String of the current TrieNode.
+         *
+         * * Iterates through all children TrieNode and join their regex String:
+         *     * if child is only a character, handle quoting
+         *     * Otherwise, retrieve the child TrieNode regex String
+         * * Add a '?' at the end if this is an end node.
+         *
+         * @return the regex String of the current TrieNode.
+         */
+        String getRegex() {
+
+            if (this.data.isEmpty()) {
+                // No data, stop here
+                return null;
+            }
+
+            // List of suffix patterns
+            final List<String> childPatterns = new ArrayList<>();
+            // List of ending characters
+            final List<Character> characters = new ArrayList<>();
+
+            for (Map.Entry<Character, TrieNode> entry : this.data.entrySet()) {
+                final String entryRegex = entry.getValue().getRegex();
+                if (entryRegex != null) {
+                    // Need to escape special / metacharacters
+                    childPatterns.add(quote(String.valueOf(entry.getKey())) + entryRegex);
+                } else {
+                    characters.add(entry.getKey());
+                }
+            }
+
+            final boolean charsOnly = childPatterns.isEmpty();
+            if (characters.size() == 1) {
+                // Need to escape special / metacharacters
+                childPatterns.add(quote(String.valueOf(characters.get(0))));
+            } else if (characters.size() > 0) {
+                // Chunking is necessary here to prevent StackOverFlow in pattern matching unions
+                final StringBuilder buf = new StringBuilder();
+                if (characters.size() < 1024) {
+                    buf.append("[");
+                    characters.forEach(character -> buf.append(quote(String.valueOf(character))));
+                    buf.append("]");
```

Comment:
I do not think you can use `Pattern.quote` inside a character class. That would be something like `[\Q?\E]` right? Which I think is a synonym for `[EQ?]` and you want just `[?]`. IIRC only `]` and `\` need to be escaped.

Reply:
Good point. Testing this and checking the `Pattern` implementation around Character classes, apparently ` / - [ ] ^ \` must be escaped.